### PR TITLE
Add implementation of psp-sysctl-policy

### PR DESCRIPTION
**Pull request by viccuad**
Example requests in *test_data/* are crafted from those in the psp-allow-privilege-escalation policy.

Implementation for https://github.com/kubewarden/policy-hub/issues/31.

### CODE REVIEW

File: validate.go
Comment (flavio) [inline validate.go:76]: I would drop/move this code somewhere else. 

I would change the internal implementation of Settings to hold:
  * `ForbiddenSysctls`
  * `ForbiddenGlobSysctls`
  * `AllowedUnsafeSysctls`

I think we could also add the `getSafeSysctls` sysctls into `AllowedUnsafeSysctls`

File: validate.go
Comment (flavio) [inline validate.go:102]: I would change the block of code with this pseudo code:

```
not_allowed = data <intesection> settings.forbidden
if len(not_allowed) > 0 {
  reject(not_allowed.join(" "))
}

to_eval = data - settings.allowed
not_allowed := []
for sysctl := range to_eval {
  if match(settings.not_allowed_glob, sysctl) {
    not_allowed = append(not_allowed, sysctl)
  }
}

if len(not_allowed) > 0 {
  reject(not_allowed.join(" "))
}

accept

```

This is just a matter of personal taste. We have Set objects with all the things that are allowed and not allowed, I just happen to like to perform "set operations" as much as possible.

File: validate.go
Comment (flavio) [inline validate.go:117]: I would drop that. Keep in mind, the go template added logging a bit too much for my taste, but just to give an idea about how to use it. Maybe we should cleanup the template too

**Review by flavio — CHANGES_REQUESTED** [review]
Overall LGTM, just some minor changes

File: README.md
Comment (ereslibre) [inline README.md:45]: I think it would be interesting to add an `allowedUnsafeSysctls` more specific than a more generic in `forbiddenSysctls` to show that it has precedence. Like:

```yaml
forbiddenSysctls:
- net.ipv6.conf.lo.*
allowedUnsafeSysctls:
- net.ipv6.conf.lo.max_addresses
```

This instance would prevent every sysctl under `net.ipv6.conf.lo`, except `net.ipv6.conf.lo.max_addresses`, that would be allowed. There might be more interesting examples than this made-up one :)

Comment (ereslibre) [reply README.md:45]: I think we should also add a test case that accepts in this situation. A pod with `sysctl` `net.ipv6.conf.lo.max_addresses` in this context should be allowed, despite `net.ipv6.conf.lo.*` is forbidden. This would test an interesting edge case we want to preserve.

File: metadata.yml
Comment (ereslibre) [inline metadata.yml:29]: I think this documentation has to specify the priority. `allowedUnsafeSysctls` takes priority over `forbiddenSysctls` if it's more specific.

**Review by ereslibre — COMMENTED** [review]
Some comments, thanks @viccuad!

File: README.md
Comment (viccuad) [reply README.md]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/b97f92b9f991fc10fa28cca6842953a66a1dfecc.

File: README.md
Comment (viccuad) [reply README.md]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/b97f92b9f991fc10fa28cca6842953a66a1dfecc.

File: README.md
Comment (viccuad) [reply README.md]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/b97f92b9f991fc10fa28cca6842953a66a1dfecc.

File: metadata.yml
Comment (viccuad) [reply metadata.yml:58]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/7ecf7bb2233c5b8cae76762c426d2fa28448942c.

File: README.md
Comment (viccuad) [reply README.md:77]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/7ecf7bb2233c5b8cae76762c426d2fa28448942c.

File: validate.go
Comment (viccuad) [reply validate.go]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/14fd3596a8ec7f57a8b6c0102ecb44fc7a55a8f0.

File: e2e.bats
Comment (viccuad) [reply e2e.bats]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/14fd3596a8ec7f57a8b6c0102ecb44fc7a55a8f0.

File: validate.go
Comment (viccuad) [reply validate.go]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/74ce514b0e87f365c2473773e7d04b22044d254e.

File: validate.go
Comment (viccuad) [reply validate.go]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/74ce514b0e87f365c2473773e7d04b22044d254e.
Opened https://github.com/kubewarden/go-policy-template/pull/8.

File: settings.go
Comment (viccuad) [reply settings.go]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/74ce514b0e87f365c2473773e7d04b22044d254e.

File: README.md
Comment (viccuad) [reply README.md]: Resolving conversations because there's a bunch (normally I would leave it to reviewers to mark as resolved).

File: metadata.yml
Comment (viccuad) [reply metadata.yml:29]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/ae703568bd79eada02a87e9f7350222ccbc6a446.

File: README.md
Comment (viccuad) [reply README.md]: Changed unit and e2e tests to check for precedence in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/ca24522f6bae4d377951496788d23b4475bd285e.
Changed readme in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/ae703568bd79eada02a87e9f7350222ccbc6a446.

File: metadata.yml
Comment (viccuad) [reply metadata.yml:81]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/520e46bcc7efcd4583086d8828ba4493159d02a2.

File: metadata.yml
Comment (viccuad) [reply metadata.yml]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/520e46bcc7efcd4583086d8828ba4493159d02a2.

File: validate.go
Comment (viccuad) [reply validate.go]: A priori sounds like a good idea at least for clarity. But then the Settings struct will not match the json (nice if we go with encoding/json). We could built an internal struct with the `forbidden`, `forbiddenGlob`, `safe` sets, instead of creating them at the beginning of evaluation. But I prefer to leave the current `knownSafeSysctls` because then We can provide a better message when rejecting.

Simplified `knownSafeSysctls` creation by using a set instead of array in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/8b03dfdf2592e8e5c2bc6b6c32ba929bde7bdb61 (I was using an array to signify that they are "constant", but it's not really needed).

I don't think it leads to a faster implementation though; right now we are iterating through the pod sysctls and if we find one sysctl that is not good, we stop and reject. With the sets, we would first perform set operations (which, given that the are implemented with maps is going to need the same iterations), and then iterate nevertheless.

File: validate.go
Comment (viccuad) [reply validate.go]: This does sound clearer, but I don't think it leads to a faster implementation though; right now we are iterating through the pod sysctls and if we find one sysctl that is not good, we stop and reject. With the sets, we would first perform set operations (which, given that the are implemented with maps is going to need the same iterations), and then iterate nevertheless, as we need to check against each blob prefix.

**Comment by viccuad** [comment]
Thanks both for the reviews! :). The CI is still not running, maybe because I set the repo to "Require approval for all outside collaborators", which sounds sane given that we have secrets for ghcr.io. I think those still need to be added, but I don't know them myself.

File: README.md
Comment (ereslibre) [inline README.md:58]:
```suggestion
    - name: net.ipv6.conf.lo.max_addresses
```

File: README.md
Comment (ereslibre) [inline README.md:79]:
```suggestion
    - name: net.ipv6.conf.lo.mtu
```

File: metadata.yml
Comment (ereslibre) [inline metadata.yml:57]:
```suggestion
        - name: net.ipv6.conf.lo.max_addresses
```

File: metadata.yml
Comment (ereslibre) [inline metadata.yml:78]:
```suggestion
        - name: net.ipv6.conf.lo.mtu
```

**Review by ereslibre — APPROVED** [review]
LGTM, just pointed out some minor typos :)

**Comment by ereslibre** [comment]
> Thanks both for the reviews! :). The CI is still not running, maybe because I set the repo to "Require approval for all outside collaborators", which sounds sane given that we have secrets for ghcr.io. I think those still need to be added, but I don't know them myself.

I think is fine to run CI here. We don't have secrets per-se, we are using `GITHUB_TOKEN` that is automatically filled by GitHub, and we use that to [sign-in in the GitHub registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry#authenticating-to-the-container-registry).

File: README.md
Comment (viccuad) [reply README.md]: doh.. fixed in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/729fcd8e9e3deeb5c0b8010ff27f55d34c4acc91.

File: README.md
Comment (viccuad) [reply README.md]: doh.. fixed in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/729fcd8e9e3deeb5c0b8010ff27f55d34c4acc91.

File: metadata.yml
Comment (viccuad) [reply metadata.yml]: doh.. fixed in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/729fcd8e9e3deeb5c0b8010ff27f55d34c4acc91.

File: metadata.yml
Comment (viccuad) [reply metadata.yml]: doh.. fixed in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/729fcd8e9e3deeb5c0b8010ff27f55d34c4acc91.

**Review by flavio — APPROVED** [review]
LGTM, thanks for the changes!

**Pull request merged**
